_posts/2021-04-22-torch-for-optimization/torch-for-optimization.Rmd

```diff
@@ -0,0 +1,640 @@
+    Value is:  0.07398106 
+    Gradient is:  -0.1603189 -0.2532476 
+
+    ...
+    ...
+
+    Iteration:  900 
+    Value is:  0.0001532408 
+    Gradient is:  -0.004811743 -0.009894371 
+
+    Iteration:  1000 
+    Value is:  6.962555e-05 
+    Gradient is:  -0.003222887 -0.006653666 
+
+While this works, it really serves to illustrate the principle only. With `torch` providing a bunch of proven optimization algorithms, there is no need for us to manually compute the candidate $\mathbf{x}$ values.
+
+## Function minimization with `torch` optimizers
+
+Instead, we let a `torch` optimizer update the candidate $\mathbf{x}$ for us. Habitually, our first try is *Adam*.
+
+### Adam (?)
+
+With Adam, optimization proceeds a lot faster. Truth be told, though, choosing a good learning rate *still* takes non-negligeable experimentation. (Try the default learning rate, 0.001, for comparison.)
+
+```{r}
+num_iterations <- 100
+
+x_star <- torch_tensor(c(-1, 1), requires_grad = TRUE)
+
+lr <- 1
+optimizer <- optim_adam(x_star, lr)
+
+for (i in 1:num_iterations) {
+  
+  if (i %% 10 == 0) cat("Iteration: ", i, "\n")
+  
+  optimizer$zero_grad()
+  value <- rosenbrock(x_star)
+  if (i %% 10 == 0) cat("Value is: ", as.numeric(value), "\n")
+  
+  value$backward()
+  optimizer$step()
+  
+  if (i %% 10 == 0) cat("Gradient is: ", as.matrix(x_star$grad), "\n\n")
+  
+}
+
+```
+
+    Iteration:  10 
+    Value is:  0.8559565 
+    Gradient is:  -1.732036 -0.5898831 
+
+    Iteration:  20 
+    Value is:  0.1282992 
+    Gradient is:  -3.22681 1.577383 
+
+    ...
+    ...
+
+    Iteration:  90 
+    Value is:  4.003079e-05 
+    Gradient is:  -0.05383469 0.02346456 
+
+    Iteration:  100 
+    Value is:  6.937736e-05 
+    Gradient is:  -0.003240437 -0.006630421 
+
+It took us about a hundred iterations to arrive at a decent value. This is a lot faster than the manual approach above, but still quite a lot. Luckily, further improvements are possible.
+
+### L-BFGS (!)
+
+Among the many `torch` optimizers commonly used in deep learning (Adam, AdamW, RMSprop ...), there is one "outsider", much better known in classic numerical optimization than in neural-networks space: L-BFGS, a.k.a. [Limited-memory BFGS](https://en.wikipedia.org/wiki/Limited-memory_BFGS), a memory-optimized implementation of the [Broyden--Fletcher--Goldfarb--Shanno optimization algorithm](https://en.wikipedia.org/wiki/Broyden%E2%80%93Fletcher%E2%80%93Goldfarb%E2%80%93Shanno_algorithm) (BFGS).
+
+BFGS is perhaps the most widely used among the so-called Quasi-Newton, second-order optimization algorithms. As opposed to the family of first-order algorithms that, in deciding on a descent direction, make use of gradient information only, second-order algorithms additionally take curvature information into account. To that end, exact Newton methods actually compute the Hessian (a costly operation), while Quasi-Newton methods avoid that cost and, instead, resort to iterative approximation.
```

Comment:
> As opposed to the family of first-order algorithms that, in deciding on a descent direction, make use of gradient information only, second-order algorithms additionally take curvature information into account.

This sentence makes sense when I get to the end of it, but it's almost as though the information is presented backwards (I don't know what you're describing—what information is taken into account with second-order algorithms—until the end).

Given this paragraph is defining technical terminology, try giving your reader simpler sentences that clearly define each component. You can even tell the reader that you're breaking down the definition/description given in that first sentence (_so-called Quasi-Newton, second-order optimization algorithm_). If someone already knew what you meant in that sentence, they'll move on. If someone (e.g. me) didn't know, they won't have to work so hard to figure it out.